File: ScanProject/JavaAnalysis/SonarJava/src/main/java/Medium/812-path-sum-iii.java
package main.java.Medium; /**https://leetcode.com/problems/path-sum-iii/ */
//Given the root of a binary tree and an integer targetSum, return the number of paths where the sum of the values along the path equals targetSum.
//The path does not need to start or end at the root or a leaf, but it must go downwards (i.e., traveling only from parent nodes to child nodes).

class TreeNode102 {
      int val;
           TreeNode102 left;
      TreeNode102 right;
      TreeNode102() {}
      TreeNode102(int val) { this.val = val; }
      TreeNode102(int val, TreeNode102 left, TreeNode102 right) {
          this.val = val;
          this.left = left;
          this.right = right;
      }
  }

class PathSumIii {
    public int pathSum(TreeNode102 root, int targetSum) {
        if(root == null){
            return 0;
        }
        return pathSumFrom(root, targetSum) + pathSum(root.left, targetSum) + pathSum(root.right, targetSum);
        
    }
    private int pathSumFrom(TreeNode102 root, int targetSum){
        if(root == null){
            return 0;
        }
        return (root.val == targetSum ? 1 : 0) + pathSumFrom(root.left, targetSum - root.val) + pathSumFrom(root.right, targetSum - root.val);
    
    }
}

    

File: ScanProject/PyAnalysis/SonarPy/Medium/812-path-sum-iii.py
#https://leetcode.com/problems/path-sum-iii/
#Given the root of a binary tree and an integer targetSum, return the number of paths where the sum of the values along the path equals targetSum.
#The path does not need to start or end at the root or a leaf, but it must go downwards (i.e., traveling only from parent nodes to child nodes).


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def pathSum(self, root, targetSum):
        """
        :type root: TreeNode
        :type targetSum: int
        :rtype: int
        """
        def dfs(root, targetSum, path):
            if not root:
                return 0
            res = 0
            path.append(root.val)
            tmp = 0
            for i in range(len(path) - 1, -1, -1):
                tmp += path[i]
                if tmp == targetSum:
                    res += 1
            res += dfs(root.left, targetSum, path)
            res += dfs(root.right, targetSum, path)
            path.pop()
            return res
        return dfs(root, targetSum, [])
    